src/Modules/ModulesGraded.jl

```diff
@@ -1,5 +1,696 @@
+  @assert length(W) == ngens(M)
+  R = base_ring(M)
+  @assert is_graded(R)
+  A = grading_group(R)
+  M.d = [W[i] * A[1] for i in 1:length(W)]
+end
+
+function degrees(M::FreeMod)
+  @assert is_graded(M)
+  return M.d
+end
+
+function degrees_of_generators(M::FreeMod)
+  return degrees(M)
+end
+
+
+function is_graded(M::FreeMod)
+  return isa(M.d, Vector{GrpAbFinGenElem})
+end
+
+###############################################################################
+# Graded Free Modules functions
+###############################################################################
+
+# for isomorphism of graded free modules, perhaps exists somewhere
+function compare_vectors_as_multisets(v1::Vector{T}, v2::Vector{T}) where {T<:Any}
+  v1_dict = Dict{T,Int}()
+  for entry in v1
+    if haskey(v1_dict, entry)
+      v1_dict[entry] += 1
+    else
+      v1_dict[entry] = 1
+    end
+  end
+  v2_dict = Dict{T,Int}()
+  for entry in v2
+    if haskey(v2_dict, entry)
+      v2_dict[entry] += 1
+    else
+      v2_dict[entry] = 1
+    end
+  end
+  return v1_dict == v2_dict
+end
+
+
+function swap!(A, i, j)
+  A[i], A[j] = A[j], A[i]
+end
+
+function generate(k, A)
+  if k == 1
+    return [copy(A)]
+  else
+    perms = generate(k - 1, A)
+    for i in 0:(k - 2)
+      if k % 2 == 0
+        swap!(A, i + 1, k)
+      else
+        swap!(A, 1, k)
+      end
+      perms = vcat(perms, generate(k - 1, A))
+    end
+    return perms
+  end
+end
+
+function permutations(v::Vector{T}) where T
+  return generate(length(v), v)
+end
+
+
+function find_bijections(v_dict, w_dict, v_key, bijections, current_bijection)
+  if v_key > length(keys(v_dict))
+    push!(bijections, deepcopy(current_bijection))
+    return nothing
+  end
+  element = collect(keys(v_dict))[v_key]
+  v_indices = v_dict[element]
+  w_indices = w_dict[element]
+  if length(v_indices) == length(w_indices)
+    for w_perm in permutations(w_indices)
+      next_bijection = deepcopy(current_bijection)
+      for (i, j) in zip(v_indices, w_perm)
+        next_bijection[i] = j
+      end
+      find_bijections(v_dict, w_dict, v_key + 1, bijections, next_bijection)
+    end
+  end
+end
+
+function get_multiset_bijection(v::Vector{T}, w::Vector{T}, all_bijections::Bool=false) where {T<:Any}
+  v_dict = Dict{T,Vector{Int}}()
+  w_dict = Dict{T,Vector{Int}}()
+  for (i, x) in enumerate(v)
+    push!(get!(v_dict, x, []), i)
+  end
+  for (i, x) in enumerate(w)
+    push!(get!(w_dict, x, []), i)
+  end
+  bijections = []
+  find_bijections(v_dict, w_dict, 1, bijections, Dict{Int,Int}())
+  return all_bijections ? bijections : (isempty(bijections) ? nothing : bijections[1])
+end
+
+###############################################################################
+# Graded Free Module elements functions
+###############################################################################
+
+function degree(el::FreeModElem)
+  !is_graded(parent(el)) && error("The parent module is not graded.")
```

Comment:
```suggestion
  @req is_graded(parent(el)) "The parent module is not graded"
```
`@req` can also be used in other places.

The majority of the error messages in `OSCAR` do not end with a period. Indeed, I had the impression that there was an agreement for this at some point.

Reply:
Yes, doc tests still need an update. There is still an issue with compact printing of polynomial rings, which should be resolved.